src/middlewared/middlewared/plugins/sysdataset.py

```diff
@@ -122,36 +120,57 @@ def sysdataset_path(self, expected_datasetname=None):
 
     @private
     async def config_extend(self, config):
-
         # Treat empty system dataset pool as boot pool
         config['pool_set'] = bool(config['pool'])
         config['pool'] = self.force_pool or config['pool'] or await self.middleware.call('boot.pool_name')
 
         config['basename'] = f'{config["pool"]}/.system'
 
         # Make `uuid` point to the uuid of current node
-        config['uuid_a'] = config['uuid']
-        is_enterprise = await self.middleware.call('system.is_enterprise')
-        is_node_b = False
-        if is_enterprise:
+        uuid_key = 'uuid'
+        if await self.middleware.call('system.is_enterprise'):
             if await self.middleware.call('failover.node') == 'B':
-                is_node_b = True
+                uuid_key = 'uuid_b'
                 config['uuid'] = config['uuid_b']
 
+        del config['uuid_b']
+
         if not config['uuid']:
             config['uuid'] = uuid.uuid4().hex
-            if is_enterprise and is_node_b:
-                attr = 'uuid_b'
-                config[attr] = config['uuid']
-            else:
-                attr = 'uuid'
             await self.middleware.call(
-                'datastore.update', 'system.systemdataset', config['id'], {f'sys_{attr}': config['uuid']}
+                'datastore.update', 'system.systemdataset', config['id'], {uuid_key: config['uuid']}, {'prefix': 'sys_'}
             )
 
         config['path'] = await self.middleware.run_in_thread(self.sysdataset_path, config['basename'])
         return config
 
+    @private
+    async def ensure_standby_uuid(self):
+        remote_uuid_key = 'uuid_b'
+        if await self.middleware.call('failover.node') == 'B':
+            remote_uuid_key = 'uuid'
+
+        local_config = await self.middleware.call('datastore.config', 'system.systemdataset', {'prefix': 'sys_'})
+        if local_config[remote_uuid_key]:
+            self.logger.debug('We already know the standby controller system dataset UUID')
+            return
+
+        remote_config = await self.middleware.call(
```

Comment:
This could crash since it's making a call to the remote system.

Reply:
Yes, but it's a part of `sync_to_peer` which will fail anyway if the remote node is down